File: databinding/testSrc/com/android/tools/idea/databinding/GeneratedCodeMatchTest.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.databinding;

import static com.android.SdkConstants.ANDROIDX_DATA_BINDING_LIB_ARTIFACT;
import static com.android.SdkConstants.DATA_BINDING_LIB_ARTIFACT;
import static com.android.tools.idea.databinding.TestDataPaths.PROJECT_WITH_DATA_BINDING_SUPPORT;
import static com.android.tools.idea.databinding.TestDataPaths.PROJECT_WITH_DATA_BINDING_ANDROID_X;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import static org.junit.runners.Parameterized.Parameters;

import com.android.builder.model.AndroidLibrary;
import com.android.ide.common.blame.Message;
import com.android.tools.idea.gradle.project.build.invoker.GradleInvocationResult;
import com.android.tools.idea.gradle.project.model.AndroidModuleModel;
import com.android.tools.idea.gradle.project.sync.GradleSyncState;
import com.android.tools.idea.res.ResourceRepositoryManager;
import com.android.tools.idea.testing.AndroidGradleProjectRule;
import com.google.common.base.Joiner;
import com.google.common.base.Predicate;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.psi.JavaPsiFacade;
import com.intellij.psi.PsiArrayType;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiClassType;
import com.intellij.psi.PsiField;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiModifier;
import com.intellij.psi.PsiModifierListOwner;
import com.intellij.psi.PsiParameter;
import com.intellij.psi.PsiType;
import com.intellij.testFramework.EdtRule;
import com.intellij.testFramework.RunsInEdt;
import com.intellij.util.containers.ContainerUtil;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.jar.JarFile;
import java.util.stream.Collectors;
import java.util.zip.ZipEntry;
import org.apache.commons.io.FileUtils;
import org.jetbrains.annotations.NotNull;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.FieldVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;

/**
 * This class compiles a real project with data binding then checks whether the generated Binding classes match the virtual ones.
 */
@RunWith(Parameterized.class)
public class GeneratedCodeMatchTest {
  public static final class TestParameters {
    @NotNull private final DataBindingMode mode;
    @NotNull private final String projectName;
    @NotNull private final String dataBindingComponentClassName;
    @NotNull private final String dataBindingLibArtifact;
    @NotNull private final String dataBindingBaseBindingClass;

    public TestParameters(@NotNull DataBindingMode mode) {
      this.mode = mode;

      projectName = mode == DataBindingMode.ANDROIDX ? PROJECT_WITH_DATA_BINDING_ANDROID_X : PROJECT_WITH_DATA_BINDING_SUPPORT;
      dataBindingComponentClassName = mode.dataBindingComponent.replace(".", "/");
      dataBindingLibArtifact = mode == DataBindingMode.ANDROIDX ? ANDROIDX_DATA_BINDING_LIB_ARTIFACT : DATA_BINDING_LIB_ARTIFACT;
      dataBindingBaseBindingClass = mode.viewDataBinding.replace(".", "/") + ".class";
    }

    @Override
    public String toString() {
      return mode.toString();
    }
  }

  @Rule
  public final AndroidGradleProjectRule myProjectRule = new AndroidGradleProjectRule();

  @Rule
  public final EdtRule myEdtRule = new EdtRule();

  @NotNull
  private final TestParameters myParameters;

  @Parameters(name = "{0}")
  public static List<TestParameters> getParameters() {
    return Lists.newArrayList(new TestParameters(DataBindingMode.SUPPORT), new TestParameters(DataBindingMode.ANDROIDX));
  }

  public GeneratedCodeMatchTest(@NotNull TestParameters parameters) {
    myParameters = parameters;
  }

  @Before
  public void setUp() {
    myProjectRule.getFixture().setTestDataPath(TestDataPaths.TEST_DATA_ROOT);
    myProjectRule.load(myParameters.projectName);
  }

  @NotNull
  private ClassReader findViewDataBindingClass() throws IOException {
    File classes = null;

    AndroidModuleModel model = AndroidModuleModel.get(myProjectRule.getAndroidFacet());
    for (AndroidLibrary lib : model.getMainArtifact().getDependencies().getLibraries()) {
      if (lib.getName().startsWith(myParameters.dataBindingLibArtifact)) {
        classes = lib.getJarFile();
      }
    }
    assert classes != null;
    assertTrue(classes.exists());

    try (JarFile classesJar = new JarFile(classes, true)) {
      ZipEntry entry = classesJar.getEntry(myParameters.dataBindingBaseBindingClass);
      assertNotNull(entry);
      return new ClassReader(classesJar.getInputStream(entry));
    }
  }

  @Test
  @RunsInEdt
  public void testGeneratedCodeMatchesExpected() throws Exception {
    // temporary fix until test model can detect dependencies properly
    GradleInvocationResult assembleDebug = myProjectRule.invokeTasks(myProjectRule.getProject(), "assembleDebug");

    assertTrue(StringUtil.join(assembleDebug.getCompilerMessages(Message.Kind.ERROR), "\n"), assembleDebug.isBuildSuccessful());

    GradleSyncState syncState = GradleSyncState.getInstance(myProjectRule.getProject());
    assertFalse(syncState.isSyncNeeded().toBoolean());
    assertEquals(ModuleDataBinding.getInstance(myProjectRule.getAndroidFacet()).getDataBindingMode(), myParameters.mode);

    // trigger initialization
    ResourceRepositoryManager.getModuleResources(myProjectRule.getAndroidFacet());

    File classesOut = new File(myProjectRule.getProject().getBasePath(), "/app/build/intermediates/javac//debug/compileDebugJavaWithJavac/classes");
    //noinspection unchecked
    Collection<File> classes = FileUtils.listFiles(classesOut, new String[]{"class"}, true);
    assertTrue("if we cannot find any class, something is wrong with the test", classes.size() > 0);
    ClassReader viewDataBindingClass = findViewDataBindingClass();

    Set<String> baseClassInfo = collectDescriptionSet(viewDataBindingClass, new HashSet<>());

    JavaPsiFacade javaPsiFacade = JavaPsiFacade.getInstance(myProjectRule.getProject());
    Set<String> missingClasses = new HashSet<>();

    Map<String, ClassReader> klassMap = classes.stream().map((file) -> {
      try {
        return new ClassReader(FileUtils.readFileToByteArray(file));
      }
      catch (IOException e) {
        e.printStackTrace();
        fail(e.getMessage());
      }
      return null;
    }).filter(kls -> kls != null)
      .collect(Collectors.toMap(
        kls -> kls.getClassName(),
        kls -> kls
      ));

    int verifiedClassCount = 0;
    for (ClassReader classReader : klassMap.values()) {
      if (!shouldVerify(viewDataBindingClass, classReader)) {
        continue;
      }
      verifiedClassCount++;
      String className = classReader.getClassName();
      PsiClass psiClass = javaPsiFacade
        .findClass(className.replace("/", "."),
                   myProjectRule.getAndroidFacet().getModule().getModuleWithDependenciesAndLibrariesScope(false));
      if (psiClass == null) {
        missingClasses.add(className);
        continue;
      }
      assertNotNull(psiClass);
      String asmInfo = collectDescriptions(classReader, baseClassInfo);
      String psiInfo = collectDescriptions(psiClass);
      assertEquals(className, asmInfo, psiInfo);
    }
    assertTrue("test sanity, should be able to find some data binding generated classes", verifiedClassCount > 3);
    assertEquals("These classes are missing", "", StringUtil.join(missingClasses, "\n"));
  }

  private boolean shouldVerify(ClassReader viewDataBindingClass, ClassReader classReader) {
    return classReader != null &&
           (viewDataBindingClass.getClassName().equals(classReader.getSuperName()) ||
            myParameters.dataBindingComponentClassName.equals(classReader.getClassName()));
  }

  @NotNull
  private static TreeSet<String> collectDescriptionSet(@NotNull ClassReader classReader, @NotNull final Set<String> exclude) {
    final TreeSet<String> set = new TreeSet<>();
    classReader.accept(new ClassVisitor(Opcodes.ASM5) {
      @Override
      public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
        List<String> interfaceList = new ArrayList<>();
        Collections.addAll(interfaceList, interfaces);
        Collections.sort(interfaceList);
        set.add(name + " : " + superName + " -> " + StringUtil.join(interfaceList, ", "));
      }

      @Override
      public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
        String info = modifierDescription(access) + " " + name + " : " + desc;
        if ((access & Opcodes.ACC_PUBLIC) != 0 && !name.startsWith("<") && !exclude.contains(info)) {
          set.add(info);
        }
        return super.visitMethod(access, name, desc, signature, exceptions);
      }

      @Override
      public FieldVisitor visitField(int access, String name, String desc, String signature, Object value) {
        String info = modifierDescription(access) + " " + name + " : " + desc;
        if ((access & Opcodes.ACC_PUBLIC) != 0 && !exclude.contains(info)) {
          set.add(info);
        }
        return super.visitField(access, name, desc, signature, value);
      }
    }, 0);
    return set;
  }

  @NotNull
  private static String collectDescriptions(@NotNull ClassReader classReader, @NotNull Set<String> exclude) {
    return StringUtil.join(collectDescriptionSet(classReader, exclude), "\n");
  }

  @NotNull
  private static String collectDescriptions(@NotNull PsiClass psiClass) {
    final TreeSet<String> set = new TreeSet<>();
    PsiClassType[] superTypes = psiClass.getSuperTypes();
    String superType = superTypes.length == 0 ? "java/lang/Object" : superTypes[0].getCanonicalText().replace('.', '/');
    List<String> sortedInterfaces = Arrays.stream(psiClass.getInterfaces()).map((klass) -> klass.getQualifiedName().replace('.', '/'))
      .sorted().collect(Collectors.toList());
    set.add(psiClass.getQualifiedName().replace('.', '/') + " : " + superType + " -> " + StringUtil.join(sortedInterfaces, ", "));
    for (PsiMethod method : psiClass.getMethods()) {
      if (method.getModifierList().hasModifierProperty(PsiModifier.PUBLIC)) {
        set.add(modifierDescription(method) + " " + method.getName() + " : " + createMethodDescription(method));
      }
    }
    for (PsiField field : psiClass.getFields()) {
      if (field.getModifierList() != null && field.getModifierList().hasModifierProperty(PsiModifier.PUBLIC)) {
        set.add(modifierDescription(field) + " " + field.getName() + " : " + createFieldDescription(field));
      }
    }
    return StringUtil.join(set, "\n");
  }

  @NotNull
  private static String createFieldDescription(@NotNull PsiField field) {
    return typeToAsm(field.getType());
  }

  @NotNull
  private static String createMethodDescription(@NotNull PsiMethod psiMethod) {
    StringBuilder res = new StringBuilder();
    PsiType returnType = psiMethod.getReturnType();
    assertNotNull(returnType);
    res.append("(");
    for (PsiParameter param : psiMethod.getParameterList().getParameters()) {
      res.append(typeToAsm(param.getType()));
    }
    res.append(")").append(typeToAsm(returnType));
    return res.toString();
  }

  @NotNull
  private static String typeToAsm(@NotNull PsiType psiType) {
    String conv = BASIC_ASM_TYPES.get(psiType);
    if (conv != null) {
      return conv;
    }
    if (psiType instanceof PsiArrayType) {
      PsiArrayType arrType = (PsiArrayType)psiType;
      return "[" + typeToAsm(arrType.getComponentType());
    }
    else {
      return "L" + psiType.getCanonicalText().replace(".", "/") + ";";
    }
  }

  @NotNull
  protected static String modifierDescription(@NotNull final PsiModifierListOwner owner) {
    return Joiner.on(" ").join(Iterables.filter(IMPORTANT_MODIFIERS.keySet(), new Predicate<String>() {
      @Override
      public boolean apply(String modifier) {
        //noinspection ConstantConditions
        return owner.getModifierList().hasModifierProperty(modifier);
      }
    }));
  }

  @NotNull
  protected static String modifierDescription(final int access) {
    return Joiner.on(" ").join(Iterables.filter(IMPORTANT_MODIFIERS.keySet(), new Predicate<String>() {
      @Override
      public boolean apply(String modifier) {
        //noinspection ConstantConditions
        return (IMPORTANT_MODIFIERS.get(modifier).intValue() & access) != 0;
      }
    }));
  }

  static Map<String, Integer> IMPORTANT_MODIFIERS = new ContainerUtil.ImmutableMapBuilder<String, Integer>()
    .put(PsiModifier.PUBLIC, Opcodes.ACC_PUBLIC)
    .put(PsiModifier.STATIC, Opcodes.ACC_STATIC)
    .build();

  static Map<PsiType, String> BASIC_ASM_TYPES =
    new ContainerUtil.ImmutableMapBuilder<PsiType, String>()
      .put(PsiType.VOID, "V")
      .put(PsiType.BOOLEAN, "Z")
      .put(PsiType.CHAR, "C")
      .put(PsiType.BYTE, "B")
      .put(PsiType.SHORT, "S")
      .put(PsiType.INT, "I")
      .put(PsiType.FLOAT, "F")
      .put(PsiType.LONG, "J")
      .put(PsiType.DOUBLE, "d").build();
}

File: databinding/testSrc/com/android/tools/idea/databinding/integration/gradle/GeneratedCodeMatchTest.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.databinding.integration.gradle

import com.android.SdkConstants.ANDROIDX_DATA_BINDING_LIB_ARTIFACT
import com.android.SdkConstants.DATA_BINDING_LIB_ARTIFACT
import com.android.tools.idea.databinding.DataBindingMode
import com.android.tools.idea.databinding.TestDataPaths
import com.android.tools.idea.databinding.TestDataPaths.PROJECT_WITH_DATA_BINDING_ANDROID_X
import com.android.tools.idea.databinding.TestDataPaths.PROJECT_WITH_DATA_BINDING_SUPPORT
import com.android.tools.idea.databinding.module.LayoutBindingModuleCache
import com.android.tools.idea.gradle.project.model.GradleAndroidModel
import com.android.tools.idea.gradle.project.sync.GradleSyncState
import com.android.tools.idea.res.ResourceRepositoryManager
import com.android.tools.idea.testing.AndroidGradleProjectRule
import com.android.tools.idea.testing.findClass
import com.google.common.collect.Lists
import com.google.common.truth.Truth.assertThat
import com.google.common.truth.Truth.assertWithMessage
import com.intellij.psi.PsiArrayType
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiField
import com.intellij.psi.PsiMethod
import com.intellij.psi.PsiModifier
import com.intellij.psi.PsiModifierListOwner
import com.intellij.psi.PsiType
import com.intellij.psi.PsiTypes
import com.intellij.testFramework.EdtRule
import com.intellij.testFramework.RunsInEdt
import com.intellij.testFramework.fixtures.JavaCodeInsightTestFixture
import org.apache.commons.io.FileUtils
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters
import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.FieldVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes
import java.io.File
import java.io.IOException
import java.util.TreeSet
import java.util.jar.JarFile

private fun String.pkgToPath(): String {
  return this.replace(".", "/")
}
private fun String.pathToPkg(): String {
  return this.replace("/", ".")
}

/**
 * This test uses both ASM (to walk through actual, compiled byte code) and PSI (to walk through
 * IntelliJ's view of a project). The following utility class helps convert both into a collection
 * of human readable descriptions that uniquely summarize a class, and allow comparing both of them
 * together.
 */
private object ClassDescriber {
  private val IMPORTANT_MODIFIERS: Map<String, Int> = mapOf(
    PsiModifier.PUBLIC to Opcodes.ACC_PUBLIC,
    PsiModifier.STATIC to Opcodes.ACC_STATIC)

  private val BASIC_ASM_TYPES: Map<PsiType, String> = mapOf(
    PsiTypes.voidType() to "V",
    PsiTypes.booleanType() to "Z",
    PsiTypes.charType() to "C",
    PsiTypes.byteType() to "B",
    PsiTypes.shortType() to "S",
    PsiTypes.intType() to "I",
    PsiTypes.floatType() to "F",
    PsiTypes.longType() to "J",
    PsiTypes.doubleType() to "d")

  private fun PsiType.toAsm(): String {
    return BASIC_ASM_TYPES[this] ?: if (this is PsiArrayType) {
      "[" + this.componentType.toAsm()
    }
    else {
      "L" + this.canonicalText.pkgToPath() + ";"
    }
  }

  /**
   * Given an ASM class reader and a list of excluded methods and fields, return a ordered set of
   * descriptions for that class.
   */
  fun collectDescriptionSet(classReader: ClassReader, exclude: Set<String> = setOf()): Set<String> {
    val descriptionSet = TreeSet<String>()

    classReader.accept(object : ClassVisitor(Opcodes.ASM5) {
      override fun visit(version: Int, access: Int, name: String, signature: String?, superName: String?, interfaces: Array<String>?) {
        val interfaceList = interfaces!!.toMutableList().apply { sort() }
        descriptionSet.add("$name : $superName -> ${interfaceList.joinToString(", ")}")
      }

      override fun visitMethod(access: Int, name: String, desc: String, signature: String?, exceptions: Array<String>?): MethodVisitor? {
        if (access and Opcodes.ACC_PUBLIC != 0 && !name.startsWith("<")) {
          descriptionSet.add("${modifierDesc(access)} $name : $desc")
        }
        return super.visitMethod(access, name, desc, signature, exceptions)
      }

      override fun visitField(access: Int, name: String, desc: String, signature: String?, value: Any?): FieldVisitor? {
        if (access and Opcodes.ACC_PUBLIC != 0) {
          descriptionSet.add("${modifierDesc(access)} $name : $desc")
        }
        return super.visitField(access, name, desc, signature, value)
      }
    }, 0)

    descriptionSet.removeAll(exclude)
    return descriptionSet
  }

  /**
   * Given PSI class, return a list of descriptions for that class.
   */
  fun collectDescriptionSet(psiClass: PsiClass): TreeSet<String> {
    val descriptionSet = TreeSet<String>()
    val superTypes = psiClass.superTypes
    val superType = if (superTypes.isEmpty()) "java/lang/Object" else superTypes[0].canonicalText.pkgToPath()
    val sortedInterfaces = psiClass.interfaces
      .map { psiInterface -> psiInterface.qualifiedName!!.pkgToPath() }
      .sorted()

    descriptionSet.add("${psiClass.qualifiedName!!.pkgToPath()} : $superType -> ${sortedInterfaces.joinToString(", ")}")
    for (method in psiClass.methods) {
      if (method.modifierList.hasModifierProperty(PsiModifier.PUBLIC)) {
        descriptionSet.add("${method.modifierDesc()} ${method.name} : ${method.methodDesc()}")
      }
    }
    for (field in psiClass.fields) {
      if (field.modifierList != null && field.modifierList!!.hasModifierProperty(PsiModifier.PUBLIC)) {
        descriptionSet.add("${field.modifierDesc()} ${field.name} : ${field.fieldDesc()}")
      }
    }

    return descriptionSet
  }

  private fun PsiField.fieldDesc() = type.toAsm()

  private fun PsiMethod.methodDesc(): String {
    val res = StringBuilder()
    val returnType = returnType!!
    res.append("(")
    for (param in parameterList.parameters) {
      res.append(param.type.toAsm())
    }
    res.append(")").append(returnType.toAsm())
    return res.toString()
  }

  private fun PsiModifierListOwner.modifierDesc(): String {
    return IMPORTANT_MODIFIERS.keys.filter { modifier -> this.modifierList!!.hasModifierProperty(modifier) }.joinToString(" ")
  }

  private fun modifierDesc(access: Int): String {
    return IMPORTANT_MODIFIERS.entries.filter { it.value and access != 0 }.joinToString(" ") { it.key }
  }
}

/**
 * This class compiles a real project with data binding then checks whether the generated Binding classes match the virtual ones.
 */
@RunWith(Parameterized::class)
class GeneratedCodeMatchTest(private val parameters: TestParameters) {
  companion object {
    @get:Parameters(name = "{0}")
    @get:JvmStatic
    val parameters: List<TestParameters>
      get() = Lists.newArrayList(TestParameters(DataBindingMode.SUPPORT), TestParameters(DataBindingMode.ANDROIDX))
  }

  private val projectRule = AndroidGradleProjectRule()

  @get:Rule
  val ruleChain = RuleChain.outerRule(projectRule).around(EdtRule())!!

  private val fixture
    get() = projectRule.fixture as JavaCodeInsightTestFixture

  class TestParameters(val mode: DataBindingMode) {
    val projectName: String =
      if (mode == DataBindingMode.ANDROIDX) PROJECT_WITH_DATA_BINDING_ANDROID_X else PROJECT_WITH_DATA_BINDING_SUPPORT
    val dataBindingLibArtifact: String =
      if (mode == DataBindingMode.ANDROIDX) ANDROIDX_DATA_BINDING_LIB_ARTIFACT else DATA_BINDING_LIB_ARTIFACT
    val dataBindingBaseBindingClass: String = mode.viewDataBinding.pkgToPath() + ".class"

    // `toString` output is used by JUnit for parameterized test names, so keep it simple
    override fun toString(): String {
      return mode.toString()
    }
  }

  @Before
  fun setUp() {
    projectRule.fixture.testDataPath = TestDataPaths.TEST_DATA_ROOT
    projectRule.load(parameters.projectName)
  }

  private fun findViewDataBindingClass(): ClassReader {
    val model = GradleAndroidModel.get(projectRule.androidFacet(":app"))!!
    val classJar = model.mainArtifact.compileClasspath.androidLibraries.first { lib ->
      lib.target.artifactAddress.startsWith(parameters.dataBindingLibArtifact)
    }.target.runtimeJarFiles.find {
      it.name == "classes.jar"
    }

    assertThat(classJar?.exists()).isTrue()
    JarFile(classJar, true).use {
      val entry = it.getEntry(parameters.dataBindingBaseBindingClass)!!
      return ClassReader(it.getInputStream(entry))
    }
  }

  @Test
  @RunsInEdt
  fun testGeneratedCodeMatchesExpected() {
    // temporary fix until test model can detect dependencies properly
    val assembleDebug = projectRule.invokeTasks("assembleDebug")
    assertThat(assembleDebug.isBuildSuccessful).isTrue()

    val syncState = GradleSyncState.getInstance(projectRule.project)
    assertThat(syncState.isSyncNeeded().toBoolean()).isFalse()
    assertThat(parameters.mode).isEqualTo(LayoutBindingModuleCache.getInstance(projectRule.androidFacet(":app")).dataBindingMode)

    // trigger initialization
    ResourceRepositoryManager.getModuleResources(projectRule.androidFacet(":app"))

    val classesOut = File(projectRule.project.basePath, "/app/build/intermediates/javac//debug/classes")

    val classes = FileUtils.listFiles(classesOut, arrayOf("class"), true)
    assertWithMessage("No compiled classes found. Something is wrong with this test.")
      .that(classes).isNotEmpty()

    val viewDataBindingClass = findViewDataBindingClass()

    // Grab a description set for the ViewDataBinding base class, as we'll strip it out of the
    // description set for the Binding subclasses, since otherwise it's just noise to us
    val baseClassInfo = ClassDescriber.collectDescriptionSet(viewDataBindingClass)

    val classMap = classes.mapNotNull<File, ClassReader> { file ->
      try {
        ClassReader(FileUtils.readFileToByteArray(file))
      }
      catch (e: IOException) {
        e.printStackTrace()
        fail(e.message)
        null
      }
    }.map { classReader -> classReader.className to classReader }.toMap()

    val context = fixture.findClass("com.android.example.appwithdatabinding.MainActivity")

    // The data binding compiler generates a bunch of stuff we don't care about in Studio. The
    // following set is what we want to make sure we generate PSI for.
    val interestingClasses = setOf(
      "${parameters.mode.packageName}DataBindingComponent",
      "com.android.example.appwithdatabinding.BR",
      "com.android.example.appwithdatabinding.databinding.ActivityMainBinding",
      "com.android.example.appwithdatabinding.databinding.MultiConfigLayoutBinding",
      "com.android.example.appwithdatabinding.databinding.MultiConfigLayoutBindingImpl",
      "com.android.example.appwithdatabinding.databinding.MultiConfigLayoutBindingLandImpl",
      "com.android.example.appwithdatabinding.databinding.NoVariableLayoutBinding"
    )
    val generatedClasses = mutableSetOf<String>()
    val missingClasses = mutableSetOf<String>()
    for (classReader in classMap.values) {
      val className = classReader.className.pathToPkg()
      if (!interestingClasses.contains(className)) {
        continue
      }
      generatedClasses.add(className)
      val psiClass = fixture.findClass(className, context)
      if (psiClass == null) {
        missingClasses.add(className)
        continue
      }

      // Convert Asm and PSI classes into description sets and verify they're the same
      val asmInfo = ClassDescriber.collectDescriptionSet(classReader, baseClassInfo)
      val psiInfo = ClassDescriber.collectDescriptionSet(psiClass)

      assertWithMessage(className).that(psiInfo).isEqualTo(asmInfo)
    }
    assertWithMessage("Failed to find expected generated data binding classes; did the compiler change?")
      .that(generatedClasses).containsExactlyElementsIn(interestingClasses)

    assertWithMessage("PSI could not be found for some generated code: ${missingClasses.joinToString(", ")}")
      .that(missingClasses).isEmpty()
  }
}
